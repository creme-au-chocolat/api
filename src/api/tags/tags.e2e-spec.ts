jest.mock('./tags.service');

import { faker } from '@faker-js/faker';
import { INestApplication } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { createTestingApp, testRequests } from '../../../test/utils/e2e';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';
import { TagsModule } from './tags.module';
import { TagsService } from './tags.service';
import {
  mockedTags,
  TagsService as MockedTagsService,
} from './__mocks__/tags.service';

const { helpers } = faker;

describe('TagsController (e2e)', () => {
  let app: INestApplication;
  let tagsService: TagsService;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [TagsModule],
    })
      .overrideProvider(TagsService)
      .useClass(MockedTagsService)
      .overrideProvider(getModelToken(TagDocument.name))
      .useValue({})
      .compile();

    tagsService = moduleRef.get(TagsService);

    app = createTestingApp(moduleRef);
    await app.init();
  });

  beforeEach(async () => {
    faker.seed(0);
  });

  describe('/tags/list/:category (GET)', () => {
    it('sort by popularity', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists?popular=true&page=1')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('sort by name', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists?popular=false&page=1')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('throws 404 error when requested page does not exist', async () => {
      const numberOfPages = await tagsService.getPageCount(CATEGORIES.artists);

      return request(app.getHttpServer())
        .get(`/tags/list/artists?popular=false&page=${numberOfPages + 1}`)
        .expect(404);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory?popular=false&page=1',
        '/tags/list/artists?popular=false&page=0',
        '/tags/list/artists?popular=false&page=1.50',
        '/tags/list/artists?popular=false&page=notanumber',
        '/tags/list/artists?popular=false&page=-1',
      ];

      await testRequests(httpServer, badRequests, 400);
    });

    it('converts parameters types', async () => {
      const httpServer = app.getHttpServer();

      const shouldOrderByPopularity: string[] = [
        '/tags/list/artists?popular=true&page=1',
        '/tags/list/artists?popular=1&page=1',
        '/tags/list/artists?popular=anything&page=1',
      ];
      const shouldOrderByName: string[] = [
        '/tags/list/artists?popular=false&page=1',
        '/tags/list/artists?page=1',
      ];
      const shouldWorks: string[] = [
        '/tags/list/artists',
        '/tags/list/artists?page=1',
      ];

      for (const req of shouldOrderByPopularity) {
        const response = await request(httpServer)
          .get(req)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      }

      for (const req of shouldOrderByName) {
        const response = await request(httpServer)
          .get(req)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      }

      for (const req of shouldWorks) {
        await request(httpServer)
          .get(req)
          .expect(200);
      }
    });
  });

  describe('/tags/list/:category/:letter (GET)', () => {
    it('returns all tags starting with specified letter, sorted by name', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('is case insensitive', async () => {
      const lowercaseResponse = await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200);

      const uppercaseResponse = await request(app.getHttpServer())
        .get('/tags/list/artists/A')
        .expect(200);

      expect(lowercaseResponse.body).toMatchSnapshot();
      expect(uppercaseResponse.body).toMatchSnapshot();
      expect(uppercaseResponse.body).toStrictEqual(lowercaseResponse.body);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory/a',
        '/tags/list/artists/notaletter',
      ];

      await testRequests(httpServer, badRequests, 400);
    });
  });

  describe('/tags/:id/details (GET)', () => {
    it('returns requested tag', async () => {
      const randomTag = helpers.arrayElement(mockedTags);

      const response = await request(app.getHttpServer())
        .get(`/tags/${randomTag.id}/details`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns 404 error when tag does not exist', async () => {
      await request(app.getHttpServer())
        .get('/tags/999999/details')
        .expect(404);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/0/details',
        '/tags/bite/details',
        '/tags/NaN/details',
      ];
      const goodRequests = [
        `/tags/${helpers.arrayElement(mockedTags).id}/details`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  describe('/tags/search (GET)', () => {
    it('returns tags including search query', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/search?q=ips')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns tags including search query and category', async () => {
      const response = await request(app.getHttpServer())
        .get('/tags/search?q=ips&category=tags')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    it('returns empty array if nothing found', async () => {
      await request(app.getHttpServer())
        .get('/tags/search?q=impossiblestring')
        .expect(200)
        .expect([]);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/search',
        '/tags/search?q=aa',
        '/tags/search?q=aaa&page=0',
        '/tags/search?q=aaa&page=1.3',
      ];
      const goodRequests = [
        `/tags/search?q=aaa&page=1&category=artists`,
        `/tags/search?q=aaa`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  afterAll(async () => {
    await app.close();
  });
});
