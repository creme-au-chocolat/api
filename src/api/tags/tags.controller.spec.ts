jest.mock('./tags.service');

import { faker } from '@faker-js/faker';
import { CacheModule, NotFoundException } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { chunk } from 'lodash';
import { Tag } from 'src/shared/types/tag.entity';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagsController } from './tags.controller';
import { TagsService } from './tags.service';
import { mockedTags } from './__mocks__/tags.service';

const { helpers } = faker;

describe('CategoriesController', () => {
  let tagsController: TagsController;

  beforeEach(async () => {
    faker.seed(1);

    const moduleRef = await Test.createTestingModule({
      controllers: [TagsController],
      providers: [TagsService],
      imports: [CacheModule.register({ ttl: 1 })],
    }).compile();

    tagsController = moduleRef.get(TagsController);
  });

  describe('getCategories', () => {
    it('should returns available categories', () => {
      const results = ['tags', 'artists', 'characters', 'parodies', 'groups'];

      expect(tagsController.getCategories()).toStrictEqual(results);
    });
  });

  describe('getTagsByCategory', () => {
    it('get values by name in first page', async () => {
      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 1 },
        ),
      ).resolves.toMatchSnapshot();
    });

    it('get value by popularity in first page', async () => {
      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 1, popular: true },
        ),
      ).resolves.toMatchSnapshot();
    });

    it.each(chunk(Object.keys(CATEGORIES), 1))(
      `correctly filters values for all categories: %s`,
      async category => {
        const currentCategory = CATEGORIES[category];

        await expect(
          tagsController.getTagsByCategory(
            { category: currentCategory },
            { page: 1 },
          ),
        ).resolves.toMatchSnapshot();
      },
    );

    it('get values by name in any page', async () => {
      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 2 },
        ),
      ).resolves.toMatchSnapshot();
    });

    it('throws an error when requested page is above total number of pages', async () => {
      await expect(
        tagsController.getTagsByCategory(
          { category: CATEGORIES.artists },
          { page: 999999 },
        ),
      ).rejects.toThrowError(NotFoundException);
    });
  });

  describe('getTagsByCategoryAndLetter', () => {
    it('returns all values starting with letter', async () => {
      await expect(
        tagsController.getTagsByCategoryAndLetter({
          category: CATEGORIES.artists,
          letter: 'A',
        }),
      ).resolves.toMatchSnapshot();
    });

    it('returns empty array when letter does not exists', async () => {
      await expect(
        tagsController.getTagsByCategoryAndLetter({
          category: CATEGORIES.artists,
          letter: 'not a letter',
        }),
      ).resolves.toEqual([]);
    });
  });

  describe('getTagById', () => {
    let tag: Tag;

    beforeEach(() => {
      tag = helpers.arrayElement(mockedTags);
    });

    it('returns tag details', async () => {
      await expect(
        tagsController.getTagById({ id: tag.id }),
      ).resolves.toMatchSnapshot();
    });
  });

  describe('searchTag', () => {
    it('returns tags matching query', async () => {
      await expect(
        tagsController.searchTag({
          q: 'i',
          category: CATEGORIES.artists,
          page: 1,
        }),
      ).resolves.toMatchSnapshot();
    });
  });
});
