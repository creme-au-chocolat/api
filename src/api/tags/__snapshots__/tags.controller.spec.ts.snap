// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesController getTagById returns tag details 1`] = `
Object {
  "category": "groups",
  "id": 486344,
  "name": "nesciunt",
  "tagged": 75934,
  "uri": "http://primary-bassoon.com",
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: artists 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 39054,
      "name": "debitis",
      "tagged": 86502,
      "uri": "https://svelte-bit.net",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 22,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: characters 1`] = `
Object {
  "data": Array [
    Object {
      "category": "characters",
      "id": 692321,
      "name": "et",
      "tagged": 71597,
      "uri": "https://sticky-epoch.com",
    },
    Object {
      "category": "characters",
      "id": 685218,
      "name": "qui",
      "tagged": 93553,
      "uri": "https://testy-head.biz",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 18,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: groups 1`] = `
Object {
  "data": Array [
    Object {
      "category": "groups",
      "id": 302332,
      "name": "esse",
      "tagged": 99718,
      "uri": "https://villainous-atm.com",
    },
    Object {
      "category": "groups",
      "id": 670466,
      "name": "sed",
      "tagged": 44345,
      "uri": "https://electric-analytics.name",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 15,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: parodies 1`] = `
Object {
  "data": Array [
    Object {
      "category": "parodies",
      "id": 198101,
      "name": "esse",
      "tagged": 45720,
      "uri": "https://incredible-cayenne.org",
    },
    Object {
      "category": "parodies",
      "id": 421107,
      "name": "ipsum",
      "tagged": 5924,
      "uri": "https://querulous-bottom-line.name",
    },
    Object {
      "category": "parodies",
      "id": 750143,
      "name": "nobis",
      "tagged": 78331,
      "uri": "https://imaginary-airship.com",
    },
    Object {
      "category": "parodies",
      "id": 345560,
      "name": "ratione",
      "tagged": 23608,
      "uri": "http://humiliating-commotion.info",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 21,
  },
}
`;

exports[`CategoriesController getTagsByCategory correctly filters values for all categories: tags 1`] = `
Object {
  "data": Array [
    Object {
      "category": "tags",
      "id": 315515,
      "name": "a",
      "tagged": 41178,
      "uri": "https://disastrous-prince.biz",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 24,
  },
}
`;

exports[`CategoriesController getTagsByCategory get value by popularity in first page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 39054,
      "name": "debitis",
      "tagged": 86502,
      "uri": "https://svelte-bit.net",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 22,
  },
}
`;

exports[`CategoriesController getTagsByCategory get values by name in any page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 789278,
      "name": "doloribus",
      "tagged": 66063,
      "uri": "https://focused-downgrade.info",
    },
    Object {
      "category": "artists",
      "id": 699757,
      "name": "sunt",
      "tagged": 61393,
      "uri": "http://well-groomed-needle.biz",
    },
  ],
  "pagination": Object {
    "page": 2,
    "total": 22,
  },
}
`;

exports[`CategoriesController getTagsByCategory get values by name in first page 1`] = `
Object {
  "data": Array [
    Object {
      "category": "artists",
      "id": 39054,
      "name": "debitis",
      "tagged": 86502,
      "uri": "https://svelte-bit.net",
    },
  ],
  "pagination": Object {
    "page": 1,
    "total": 22,
  },
}
`;

exports[`CategoriesController getTagsByCategoryAndLetter returns all values starting with letter 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 828844,
    "name": "autem",
    "tagged": 25470,
    "uri": "https://boiling-obesity.com",
  },
]
`;

exports[`CategoriesController searchTag returns tags matching query 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 798603,
    "name": "inventore",
    "tagged": 87985,
    "uri": "https://keen-councilman.net",
  },
  Object {
    "category": "artists",
    "id": 203906,
    "name": "dignissimos",
    "tagged": 55552,
    "uri": "https://far-working.info",
  },
  Object {
    "category": "artists",
    "id": 370083,
    "name": "libero",
    "tagged": 46215,
    "uri": "http://competent-baritone.net",
  },
  Object {
    "category": "artists",
    "id": 878831,
    "name": "nisi",
    "tagged": 28147,
    "uri": "https://grimy-mozzarella.org",
  },
  Object {
    "category": "artists",
    "id": 599109,
    "name": "quibusdam",
    "tagged": 57706,
    "uri": "http://coarse-kinase.org",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "debitis",
    "tagged": 86502,
    "uri": "https://svelte-bit.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "modi",
    "tagged": 4830,
    "uri": "https://unrealistic-self-control.net",
  },
  Object {
    "category": "artists",
    "id": 789278,
    "name": "doloribus",
    "tagged": 66063,
    "uri": "https://focused-downgrade.info",
  },
  Object {
    "category": "artists",
    "id": 949015,
    "name": "perferendis",
    "tagged": 42681,
    "uri": "http://second-orient.net",
  },
]
`;
